### PR TITLE
Generate Genesis Delegation certs

### LINKED ISSUE #1042: Generate _Genesis Delegate_ Certificates

### PR DESCRIPTION
**Pull request by uroboros**
Closes #1042

### CODE REVIEW

File: byron/semantics/executable-spec/src/Control/State/Transition/Trace/Generator/QuickCheck.hs [hunk 1]
```diff
@@ -42,6 +42,7 @@ import qualified Control.State.Transition as STS
 import           Control.State.Transition.Trace (Trace)
 import qualified Control.State.Transition.Trace as Trace
 
+import qualified Debug.Trace as D
```
Comment (mgudemann) [inline byron/semantics/executable-spec/src/Control/State/Transition/Trace/Generator/QuickCheck.hs]: as discussed, maybe we should add a log/debug module that could take care of that and which could be silenced in case we don't want any output

File: shelley/chain-and-ledger/executable-spec/src/LedgerState.hs [hunk 2]
```diff
@@ -356,16 +356,22 @@ genesisCoins outs = UTxO $
...
     (Coin 0)
     emptyUpdateState)
-  emptyDelegation
+  (DPState dState emptyPState)
   0
+  where
+    dState = DState (StakeCreds Map.empty)
+                    Map.empty Map.empty Map.empty Map.empty
+                    (GenDelegs genDelegs0)
+                    Map.empty
```
Comment (mgudemann) [inline shelley/chain-and-ledger/executable-spec/src/LedgerState.hs]: there'a an `emptyDState` in `LedgerState` that you could use and just overwrite the record entry for genesis delegation. I think that would be more readable

File: shelley/chain-and-ledger/executable-spec/test/Generator/Core.hs [hunk 3]
```diff
@@ -24,6 +24,7 @@ module Generator.Core
 
 import           Cardano.Crypto.VRF (deriveVerKeyVRF, genKeyVRF)
 import           Crypto.Random (drgNewTest, withDRG)
+import qualified Data.Map.Strict as Map
```
Comment (mgudemann) [inline shelley/chain-and-ledger/executable-spec/test/Generator/Core.hs]: better add explicit imports

File: shelley/chain-and-ledger/executable-spec/test/Generator/Core/QuickCheck.hs [hunk 4]
```diff
@@ -66,6 +68,16 @@ mkKeyPairs n
 traceKeyPairs :: KeyPairs
 traceKeyPairs = mkKeyPairs <$> [1 .. 150]
 
+numCoreNodes :: Word64
+numCoreNodes = 7
+
+coreNodes :: [(CoreKeyPair, KeyPair)]
+coreNodes = [ ( (toKeyPair . mkGenKey) (x, 0, 0, 0, 0)
+              , (toKeyPair . mkKeyPair) (x, 0, 0, 0, 1))
+            | x <- [1001..1000+numCoreNodes]]
```
Comment (mgudemann) [inline shelley/chain-and-ledger/executable-spec/test/Generator/Core/QuickCheck.hs]: would you mind adding a comment to this that includes the reasoning for this magic number?

File: shelley/chain-and-ledger/executable-spec/test/Generator/LedgerTrace/QuickCheck.hs [hunk 5]
```diff
@@ -32,7 +32,8 @@ instance TQC.HasTrace LEDGER Word64 where
               <*> genPParams
               <*> genCoin 0 1000
 
+  -- TODO pass core node keys as inputs
   sigGen _ ledgerEnv ledgerSt =
-    genTx ledgerEnv ledgerSt traceKeyPairs traceVRFKeyPairs
+    genTx ledgerEnv ledgerSt traceKeyPairs [] traceVRFKeyPairs
```
Comment (mhuesch) [inline shelley/chain-and-ledger/executable-spec/test/Generator/LedgerTrace/QuickCheck.hs]: Does this cause the QC generator to only generate empty lists of `[CoreKeyPair]`s?

File: shelley/chain-and-ledger/executable-spec/test/Rules/ClassifyTraces.hs [hunk 6]
```diff
@@ -47,6 +47,10 @@ relevantCasesAreCovered = withMaxSuccess 500 . property $ do
               (traceLength tr <= 20 * length (filter isDelegation certs_))
               "there is at least 1 Delegation certificate for every 10 transactions"
 
+     , cover_ 0 -- TODO @uroboros (currently not producing any Genesis Deleg certs)
+              (traceLength tr <= 40 * length (filter isGenesisDelegation certs_))
+              "there is at least 1 Genesis Delegation certificate for every 40 transactions"
```
Comment (mhuesch) [inline shelley/chain-and-ledger/executable-spec/test/Rules/ClassifyTraces.hs]: Is this related to the `[]` keyspace for `coreKeys`?

File: shelley/chain-and-ledger/executable-spec/test/Test/Serialization.hs [hunk 7]
```diff
@@ -124,206 +119,243 @@ getRawInstallerHash (InstallerHash hsh) = getHash hsh
...
+  , let rs = Map.singleton (KeyHashObj testKeyHash1) 77
+    in
+    checkEncoding "mir"
+    (InstantaneousRewards rs)
+    ( T (TkListLen 2
+       . TkWord 9) -- make instantaneous rewards cert
+      <> S rs
     )
  ]
-
```
Comment (mhuesch) [inline shelley/chain-and-ledger/executable-spec/test/Test/Serialization.hs]: Only glanced over this; not sure of the context.

File: shelley/chain-and-ledger/cddl-spec/shelley.cddl [hunk 8]
```diff
@@ -122,6 +138,7 @@ $signature /= bytes
 
 $vrf_keyhash /= bytes
 
+$vrf_vkey /= bytes
 $vrf_proof /= bytes
 
 $kes_vkey /= bytes
```
[the remark this answers is not in the document]
Comment (uroboros) [reply shelley/chain-and-ledger/cddl-spec/shelley.cddl]: these snuck in during a rebase, but I've snuck them out again!

File: shelley/chain-and-ledger/executable-spec/test/Generator/Core.hs (hunk 3 above)
Comment (uroboros) [reply shelley/chain-and-ledger/executable-spec/test/Generator/Core.hs]: 👍

File: shelley/chain-and-ledger/executable-spec/test/Generator/Delegation.hs [hunk 9]
```diff
@@ -79,23 +79,29 @@ genDCerts keys vrfKeys pparams dpState slot ttl = do
...
 genDCert
   :: KeyPairs
+  -> [CoreKeyPair]
   -> VrfKeyPairs
   -> PParams
   -> DPState
   -> Slot
-  -> Gen (Maybe (DCert, KeyPair))
-genDCert keys vrfKeys pparams dpState slot =
+  -> Gen (Maybe (DCert, Either KeyPair [CoreKeyPair]))
```
[the remark this answers is not in the document]
Comment (uroboros) [reply shelley/chain-and-ledger/executable-spec/test/Generator/Delegation.hs]: 👍

File: shelley/chain-and-ledger/executable-spec/test/Generator/Core.hs (hunk 3 above)
Comment (dnadales) [reply shelley/chain-and-ledger/executable-spec/test/Generator/Core.hs]: FWIW `Map` is usually used in this way. Otherwise you have to hide some operations from `Prelude`.

File: shelley/chain-and-ledger/executable-spec/test/Generator/LedgerTrace/QuickCheck.hs (hunk 5 above)
Comment (uroboros) [reply shelley/chain-and-ledger/executable-spec/test/Generator/LedgerTrace/QuickCheck.hs]: yup! good catch thanks ;)

File: byron/semantics/executable-spec/src/Control/State/Transition/Trace/Generator/QuickCheck.hs (hunk 1 above)
Comment (mhuesch) [reply byron/semantics/executable-spec/src/Control/State/Transition/Trace/Generator/QuickCheck.hs]: I happened to glance over the Haxl library today and noticed they use this [`trace_` idiom](https://github.com/facebook/Haxl/blob/1119155782233ed4e99bbf0c2a0fac9c8ee9a933/Haxl/Core/Util.hs), which might be useful for this case.

Comment (mgudemann) [reply byron/semantics/executable-spec/src/Control/State/Transition/Trace/Generator/QuickCheck.hs]: that looks good, how about we use something like `trace_` and add an additional parameter like `info` `warning` `error` for different levels of information? One probably would not want all output in the case of looking for an error
